flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSourceSinkFactoryBase.java

```diff
@@ -187,9 +190,21 @@
 	// --------------------------------------------------------------------------------------------
 
 	/**
-	 * Returns the Kafka version.
+	 * Returns a hard-coded Kafka version.
+	 *
+	 * @deprecated This method can be dropped once we only maintain the universal Kafka connector.
 	 */
-	protected abstract String kafkaVersion();
+	@Deprecated
+	protected abstract Optional<String> kafkaVersion();
+
+	/**
+	 * Returns the pattern that a validator should accept as Kafka version.
+	 */
+	protected Pattern[] kafkaVersionPattern() {
```

Comment:
The reason for two separate methods was that the previous factories explicitly matched only if the version was correct. But I agree that the logic might have been unnecessarily complicated. I removed the `kafkaVersion()` method. I think `Pattern` is still fine because is not a dependency and also not exposed as factories are a pure Flink internal thing. Once we drop the old Kafka connectors we don't need to "expose" the pattern anymore.

Reply:
I revert my comment from a above. The explicit version is still necessary since version `10` depends on version `9`, thus both factories are present in a Kafka 0.10 module. The two factories would interfere with each other if a hard-coded version is not part of the context.